### PR TITLE
Add 2015-07-28-long-paths

### PR DESCRIPTION
**Pull request by terrajobst**

### CODE REVIEW

File: 2015-07-28-long-paths/README.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
+# API Review 2015-07-28
+
+This API review was also recorded and is available on [Google+](https://plus.google.com/events/cu7i66er1ien3s3975o8a5hs8i8).
+
+## Overview
+
+In this API review we discussed how we want address long path limitations due
```
Comment (akoeplinger) [inline 2015-07-28-long-paths/README.md]: we want **to** address

File: 2015-07-28-long-paths/README.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+[Video](https://plus.google.com/events/cu7i66er1ien3s3975o8a5hs8i8)
+
+### Enabling Long Paths
+
+* .NET Core needs to handle > MAX_PATH (260) paths for Unix
+  - Enable CoreFX (Unix mostly done)
+  - Enable CoreCLR (Assembly & resource loading, not done)
+* Also want to support > MAX_PATH on Windows
+  - Sharing early thinking, will recircle with deeper dives
+  - Many details need uncovered / answered
```
Comment (akoeplinger) [inline 2015-07-28-long-paths/README.md]: need **to be** uncovered ?

File: 2015-07-28-long-paths/README.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+### Windows Extended Path Syntax
+
+* Paths start with special character sequence
+  - `\\?\` for local paths
+  - `\\?\UNC\` for UNCs (`\\Server\Share`)
+* Allows ~32K length paths
+  - exact path length has to fit when converted to internal NT path
+* Additional characters (trailing spaces, period)
+* Allows legacy device names
+* Does not normalize (`.`, `..``)
```
Comment (akoeplinger) [inline 2015-07-28-long-paths/README.md]: remove extraneous backtick at the end

File: 2015-07-28-long-paths/README.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  - Legacy character support (no trailing spaces, etc.)
+  - Extended syntax not given back
+* Intercept before P/Invoke
+  - Append extended if needed
+  - Trim extended on return if added
+
+### Allow Extended Syntax (`\\?\`)
+
+* Don't attempt to normalize
+* Very basic correctness checks (invalid chars, `.`, `..``)
```
Comment (akoeplinger) [inline 2015-07-28-long-paths/README.md]: remove extraneous backtick at the end

**Comment by akoeplinger** [comment]
Left a few minor comments, otherwise LGTM :+1:

**Comment by terrajobst** [comment]
Thanks! Fixed.

**Comment by JeremyKuhne** [comment]
LGTM

**Comment by tmat** [comment]
Nice! Yay for Path.PathsAreEquivalent(string firstPath, string secondPath)
Just curious, will it use some kind of file system specific file identity? 

Also, could variants of Path.Combine, Path.GetFileName, Path.GetDirectory, Path.GetExtension that don't check for invalid characters be provided? So that we can work with patterns like Path.GetDirectory("C:\temp\*.cs") == "C:\temp" etc.

**Comment by terrajobst** [comment]
> So that we can work with patterns like  Path.GetDirectory(@"C:\temp\*.cs") == "C:\temp"  etc.
Interesting. I assume you want this API to filter break user input apart, I guess?

**Comment by tmat** [comment]
Re comparing paths: The scenario that might break is when a case-insensitive FS partition is mounted as a subdirectory of case-sensitive FS. Then half of the path needs to be compared case-insensitively and half case-sensitively.

Correct. Roslyn has custom implementations of these Path API for that reason. We'd prefer not having them.

**Comment by terrajobst** [comment]
> The scenario that might break is when a case-insensitive FS partition is mounted as a subdirectory of case-sensitive FS. Then half of the path needs to be compared case-insensitively and half case-sensitively.

Yeah, @jaredpar brought this up in the meeting as well. Hence the proposal to expose an API that gets a canonical representation. In that, all inputs would be using the FS way to get a canonical representation. Afterwards, all comparisons could be done in a case sensitive fashion.

**Pull request merged**